ndb/src/google/cloud/ndb/model.py

```diff
@@ -2302,17 +2302,135 @@ def _db_get_value(self, v, unused_p):
 
 
 class DateProperty(DateTimeProperty):
+    """A property that contains :class:`~datetime.date` values.
+
+    .. automethod:: _to_base_type
+    .. automethod:: _from_base_type
+    .. automethod:: _validate
+    """
+
     __slots__ = ()
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    def _validate(self, value):
+        """Validate a ``value`` before setting it.
+
+        Args:
+            value (~datetime.date): The value to check.
+
+        Raises:
+            .BadValueError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
+            raise exceptions.BadValueError(
+                "Expected date, got {!r}".format(value)
+            )
+
+    def _to_base_type(self, value):
+        """Convert a value to the "base" value type for this property.
+
+        Args:
+            value (~datetime.date): The value to be converted.
+
+        Returns:
+            ~datetime.datetime: The converted value: a datetime object with the
+            time set to ``00:00``.
+
+        Raises:
+            TypeError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
```

Comment:
There [was][1] an `assert` here but it didn't make sense because

1. `_validate()` already checked the type
1. The `_date_to_datetime()` helper checks the type.

FWIW I moved the `_date_to_datetime()` implementation here wholesale

[1]: https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/cf4cab3f1f69cd04e1a9229871be466b53729f3f/ndb/model.py#L2183

Reply:
:+1: